zipline/assets/asset_writer.py

```diff
@@ -0,0 +1,744 @@
+import pandas as pd
+import numpy as np
+from pandas.tseries.tools import normalize_date
+from six import with_metaclass, string_types
+from abc import (
+    ABCMeta,
+    abstractmethod,
+)
+from zipline.errors import (
+    ConsumeAssetMetaDataError,
+    InvalidAssetType,
+    SidAssignmentError,
+)
+from zipline.assets import (
+    Asset, Equity, Future
+)
+
+ASSET_FIELDS = [
+    'sid',
+    'asset_type',
+    'symbol',
+    'asset_name',
+    'start_date',
+    'end_date',
+    'first_traded',
+    'exchange',
+    'notice_date',
+    'root_symbol',
+    'expiration_date',
+    'contract_multiplier',
+    # The following fields are for compatibility with other systems
+    'file_name',  # Used as symbol
+    'company_name',  # Used as asset_name
+    'start_date_nano',  # Used as start_date
+    'end_date_nano'  # Used as end_date
+]
+
+# Expected fields for an Asset's metadata
+ASSET_TABLE_FIELDS = [
+    'sid',
+    'symbol',
+    'asset_name',
+    'start_date',
+    'end_date',
+    'first_traded',
+    'exchange'
+]
+
+# Expected fields for an Asset's metadata
+FUTURE_TABLE_FIELDS = ASSET_TABLE_FIELDS + [
+    'root_symbol_id',
+    'notice_date',
+    'expiration_date',
+    'contract_multiplier',
+]
+
+EQUITY_TABLE_FIELDS = ASSET_TABLE_FIELDS
+
+EXCHANGE_TABLE_FIELDS = [
+    'exchange_id',
+    'exchange',
+    'timezone'
+]
+
+ROOT_SYMBOL_TABLE_FIELDS = [
+    'root_symbol_id',
+    'root_symbol',
+    'sector',
+    'description',
+    'exchange_id'
+]
+
+
+class AssetDBWriter(with_metaclass(ABCMeta)):
+    """
+    Class used to write arbitrary data to SQLite database.
+    Concrete subclasses will implement the logic for a specific
+    input datatypes by implementing the gen_data method.
+
+    Methods
+    -------
+    write_all(db_conn, fuzzy_char=None, allow_sid_assignment=True,
+              constraints=False)
+        Write the data supplied at initialization to the database.
+    init_db(db_conn, constraints=False)
+        Create the SQLite tables (called by write_all).
+    gen_data(self, equities, futures, exchanges, root_symbols)
```

Comment:
I don't think this is still the signature, we call it with no args in the body of `write_all`